### DISCUSSION TITLE
Node classification on new nodes (or expanded graph)

### DISCUSSION THREAD

**Discussion by michalisfrangos** [pyg-team/pytorch_geometric#3849] [Q&A]
Hi all, 

All the examples I have seen,  train on a subset of nodes (selected by train_mask) and test on another set of nodes of the same graph (selected by test_mask) however the structure of the graph is present during training .  Any reason why we do not remove the  "test" nodes completely  during training (i.e., we test on new/unseen nodes and connectivity)? 

Thanks,
M

**Comment by rusty1s** [comment]
You can do this, *e.g.*, by obtaining subgraphs via `data.subgraph(train_mask)`, but in general, such transductive learning helps to make better predictions (due to having access to more data). As such, in case the full graph is given to us (but only a fraction of labels), there is no reason to not make fully use of it.

Hope this clarifies your concerns. Note that the `PPI` dataset is a classic node-level prediction dataset that is inductive, i.e. we test on unseen graphs. See [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/ppi.py) for an example.

**Reply by michalisfrangos** [reply]
Thanks @rusty1s for your response and the suggested example, I'll have a look! 

Ok I see your point that at training we better use all available structure. 

I was considering in mind practical applications, e.g. when we have a new customer (in our graph of products/customers); I would have throught that performance metrics would be more realistic, if we were to ignore a part of the graph completely during training ... and see how it behaves on new nodes/structure .

**Comment by Srivathsan-Srinivas** [comment]
Thanks @michalisfrangos for asking this very practical question and thanks @rusty1s for the pointer. I have the same question (posed in a different way): I intend to track a customer over a period of time and say I have a label for him as "Type A" or "Type B". Now, a new customer comes in and I track him for a short period of time and I want to classify this customer (binary classification for now). 

I do not want to build a huge graph **at inference time** (historical graph + new node's small-graph) and then infer the class of the new node (latency may become a problem). I would like to have built a graph from historical dataset first and then, when a new node comes in, I want to be able to classify it. **What are the logical steps to achieve it?**

One approach that I found is from https://keras.io/examples/graph/gnn_citations/, (scroll all the way down in that link), the keys aspects are highlighted below:

1. add new node(s) to the existing (historical) graph -> append to the existing/historical node-features.
2. add edges from new node(s) to the set of existing edges.
3. update 1 and 2 in the model.
4. Do prediction on new node instances.

I haven't tried that above yet. Will be trying it shortly. Thought I will share it here and see if others have any alternate / easier approaches.

**Reply by rusty1s** [reply]
Yes, this sounds correct. Alternatively, you can just maintain the local subgraph around new nodes and pass that to the GNN (any information outside from what is reachable by the GNN will not be used anyway).
